// jw-angular-pagination/esm2015/lib/jw-pagination.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { JwPaginationComponent } from './jw-pagination.component';
import * as ɵngcc0 from '@angular/core';
let JwPaginationModule = class JwPaginationModule {
};
JwPaginationModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: JwPaginationModule });
JwPaginationModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function JwPaginationModule_Factory(t) { return new (t || JwPaginationModule)(); }, imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(JwPaginationModule, { declarations: function () { return [JwPaginationComponent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [JwPaginationComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(JwPaginationModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [JwPaginationComponent],
                exports: [JwPaginationComponent]
            }]
    }], null, null); })();
export { JwPaginationModule };
