// ngx-country-picker/esm5/country-picker.service.js
import * as tslib_1 from "tslib";
import { Inject, Injectable } from '@angular/core';
import { HttpClient, HttpResponse } from '@angular/common/http';
import { throwError } from 'rxjs';
import { catchError, map } from 'rxjs/operators';
import { COUNTRY_PICKER_CONFIG } from './country-picker.config';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common/http';
var CountryPickerService = /** @class */ (function () {
    function CountryPickerService(config, _http) {
        this._http = _http;
        this._data = null;
        this._baseUrl = config.baseUrl;
        this._filename = config.filename;
        this._data = this._loadData();
    }
    CountryPickerService_1 = CountryPickerService;
    CountryPickerService.handleError = function (error) {
        var errMsg;
        if (error instanceof HttpResponse) {
            if (error.status === 404) {
                errMsg = 'Error loading countries.json file.'
                    + ' Please configure WebPack and load countries.json assets to your root folder';
            }
            else {
                var body = error.body || '';
                var err = body.error || JSON.stringify(body);
                errMsg = error.status + " - " + (error.statusText || '') + " " + err;
            }
        }
        else {
            errMsg = error.message ? error.message : error.toString();
        }
        console.error(errMsg);
        return throwError(errMsg);
    };
    CountryPickerService.prototype.getCountries = function () {
        return this._data
            .pipe(map(function (countries) {
            return countries.map(function (country) {
                country.name.native[0] = country.name.native[Object.keys(country.name.native)[0]];
                return country;
            });
        }));
    };
    CountryPickerService.prototype.getBaseUrl = function () {
        return this._baseUrl;
    };
    CountryPickerService.prototype._loadData = function () {
        return this._http.get(this._baseUrl + this._filename)
            .pipe(catchError(CountryPickerService_1.handleError));
    };
    var CountryPickerService_1;
    CountryPickerService.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [COUNTRY_PICKER_CONFIG,] }] },
        { type: HttpClient }
    ]; };
    CountryPickerService = CountryPickerService_1 = tslib_1.__decorate([ tslib_1.__param(0, Inject(COUNTRY_PICKER_CONFIG))
    ], CountryPickerService);
CountryPickerService.ɵfac = function CountryPickerService_Factory(t) { return new (t || CountryPickerService)(ɵngcc0.ɵɵinject(COUNTRY_PICKER_CONFIG), ɵngcc0.ɵɵinject(ɵngcc1.HttpClient)); };
CountryPickerService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: CountryPickerService, factory: function (t) { return CountryPickerService.ɵfac(t); } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(CountryPickerService, [{
        type: Injectable
    }], function () { return [{ type: undefined, decorators: [{
                type: Inject,
                args: [COUNTRY_PICKER_CONFIG]
            }] }, { type: ɵngcc1.HttpClient }]; }, null); })();
    return CountryPickerService;
}());
export { CountryPickerService };
