// ngx-country-picker/esm2015/country-picker.component.js
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, Input, OnInit } from '@angular/core';
import { CountryPickerService } from './country-picker.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';

function CountryPickerComponent_option_1_img_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "img", 3);
} if (rf & 2) {
    const c_r1 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("src", ctx_r2.baseUrl + c_r1.cca3.toLowerCase() + ".svg", ɵngcc0.ɵɵsanitizeUrl);
} }
function CountryPickerComponent_option_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "option", 1);
    ɵngcc0.ɵɵtemplate(1, CountryPickerComponent_option_1_img_1_Template, 1, 1, "img", 2);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const c_r1 = ctx.$implicit;
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("value", ctx_r0.getValue(c_r1));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.flag);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1("", ctx_r0.getName(c_r1), " ");
} }
let CountryPickerComponent = class CountryPickerComponent {
    constructor(_countryPickerService, _cdr) {
        this._countryPickerService = _countryPickerService;
        this._cdr = _cdr;
        this.flag = false;
        this.setValue = 'cca3';
        this.setName = 'name.common';
        this.classes = ['form-control', 'form-control-sm'];
        this.countries = [];
        this.baseUrl = _countryPickerService.getBaseUrl() + 'data/';
    }
    ngOnInit() {
        this._countryPickerService.getCountries()
            .subscribe(countries => {
            this.countries = countries.sort((a, b) => {
                let na = this.getName(a);
                let nb = this.getName(b);
                if (na > nb) {
                    return 1;
                }
                if (na < nb) {
                    return -1;
                }
                return 0;
            });
            this._cdr.markForCheck();
        });
    }
    getValue(obj) {
        return this.setValue.split('.').reduce((o, i) => o[i], obj);
    }
    getName(obj) {
        return this.setName.split('.').reduce((o, i) => o[i], obj);
    }
};
CountryPickerComponent.ɵfac = function CountryPickerComponent_Factory(t) { return new (t || CountryPickerComponent)(ɵngcc0.ɵɵdirectiveInject(CountryPickerService), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef)); };
CountryPickerComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: CountryPickerComponent, selectors: [["country-picker"]], inputs: { flag: "flag", setValue: "setValue", setName: "setName", classes: "classes" }, decls: 2, vars: 3, consts: [[3, "value", 4, "ngFor", "ngForOf"], [3, "value"], [3, "src", 4, "ngIf"], [3, "src"]], template: function CountryPickerComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "select");
        ɵngcc0.ɵɵtemplate(1, CountryPickerComponent_option_1_Template, 3, 3, "option", 0);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵclassMap(ctx.classes);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.countries);
    } }, directives: [ɵngcc1.NgForOf, ɵngcc1.NgIf], encapsulation: 2, changeDetection: 0 });
CountryPickerComponent.ctorParameters = () => [
    { type: CountryPickerService },
    { type: ChangeDetectorRef }
];
tslib_1.__decorate([
    Input()
], CountryPickerComponent.prototype, "flag", void 0);
tslib_1.__decorate([
    Input()
], CountryPickerComponent.prototype, "setValue", void 0);
tslib_1.__decorate([
    Input()
], CountryPickerComponent.prototype, "setName", void 0);
tslib_1.__decorate([
    Input()
], CountryPickerComponent.prototype, "classes", void 0);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(CountryPickerComponent, [{
        type: Component,
        args: [{
                selector: 'country-picker',
                template: `
      <select [class]="classes">
          <option *ngFor="let c of countries" [value]="getValue(c)">
              <img *ngIf="flag" [src]="baseUrl + c.cca3.toLowerCase() + '.svg'">{{ getName(c) }}
          </option>
      </select>
  `,
                changeDetection: ChangeDetectionStrategy.OnPush
            }]
    }], function () { return [{ type: CountryPickerService }, { type: ɵngcc0.ChangeDetectorRef }]; }, { flag: [{
            type: Input
        }], setValue: [{
            type: Input
        }], setName: [{
            type: Input
        }], classes: [{
            type: Input
        }] }); })();
export { CountryPickerComponent };
